pkg/operator/credentialsrequest/credentialsrequest_controller.go

```diff
@@ -227,6 +227,24 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
+	// predicate functions to filter the events based on the AWS resourceTags presence.
+	infraResourcePredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return hasResourceTags(e.Object)
+		},
+		DeleteFunc: func(event.DeleteEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return areTagsUpdated(e.ObjectOld, e.ObjectNew)
+		},
+	}
+	// Watch for the changes happening to Infrastructure Resource
+	err = c.Watch(&source.Kind{Type: &configv1.Infrastructure{}}, allCredRequestsMapFn, infraResourcePredicate)
```

Comment:
While requeueing all the CredentialsRequest resources makes sense, from looking at the code flow in this controller, it will end up doing nothing because a change to the Infrastructure resource is not something that would get passed all these checks: https://github.com/openshift/cloud-credential-operator/blob/master/pkg/operator/credentialsrequest/credentialsrequest_controller.go#L529-L540

Reply:
Yes @joelddiaz , I observed this and the tags are getting updated whenever the next sync happens.
I discussed this with my team members and we have come up with two ways of resolving this problem.
1. Addition of a new field "ResourceTags" in the "CredentialsRequest" resource's status field, compare these tags with the Infrastructure's ResourceTags whenever the Reconcile function is called.
2. Addition of a new controller to monitor the Infrastructure resource, corresponding reconcile function consists of listing the CredentialaRequest resources, calling the respective actuators etc. similar to that of happening in the existing Reconcile function [here](https://github.com/openshift/cloud-credential-operator/blob/master/pkg/operator/credentialsrequest/credentialsrequest_controller.go#L358).

Please let me know your thoughts on these and also, do you have any suggestions apart from the above mentioned proposals?

Reply:
@akhil-rane Could you have a look at this once you are available?
Thanks

Reply:
@sairameshv option (1) sounds good to me. Instead of having `ResouceTags` field under `CredenetialsRequest.Status`, we should have it under `CredenetialsRequest.Status.ProviderStatus` as this field is specific to cloud provider AWS. Does that make sense?

Reply:
@akhil-rane, I think that's a good idea but the only concern is that the code is becoming very much specific to AWS (Decoding the Provider specific status, comparing with provider tags in Infra object, etc.) as pointed out by @joelddiaz here in this [comment](https://github.com/openshift/cloud-credential-operator/pull/445#discussion_r799484205). Also, in future, if we are to update the tags on the Azure resources, it would again require changes specific to Azure in the generic credentialsrequest controller code. Don't you think it's better to define the `ResourceTags` field in the `CredenetialsRequest.Status` itself?

Moreover, just FYI, the [enhancement proposal](https://github.com/openshift/enhancements/pull/1051) is not insisting on any time limit under which the tags must be updated. Hence, the existing code works fine as of now with a maximum delay of 1hr updating the tags on the AWS resources.

Reply:
> Moreover, just FYI, the https://github.com/openshift/enhancements/pull/1051 is not insisting on any time limit under which the tags must be updated. Hence, the existing code works fine as of now with a maximum delay of 1hr updating the tags on the AWS resources.

Delaying the reconcile by 1 hour would not be a good user experience so I think we should try and update tags on aws resources immediately.

> I think that's a good idea but the only concern is that the code is becoming very much specific to AWS (Decoding the Provider specific status, comparing with provider tags in Infra object, etc.) as pointed out by @joelddiaz here in this https://github.com/openshift/cloud-credential-operator/pull/445#discussion_r799484205. Also, in future, if we are to update the tags on the Azure resources, it would again require changes specific to Azure in the generic credentialsrequest controller code. Don't you think it's better to define the ResourceTags field in the CredenetialsRequest.Status itself?

I agree with you on this. This would introduce more AWS specific code in credentials_request controller. But again we cannot store Tags on Credentials request status. Tags are very specific to AWS. Azure might have tags but other cloud providers might not. Can we have a field called `LastSyncInfrastructureResourceVersion` (something similar to [this](https://github.com/openshift/cloud-credential-operator/blob/ed0612946f4b77a18aa1e631b83abef9d856697a/pkg/apis/cloudcredential/v1/types_credentialsrequest.go#L82-L87))  in the CredentialsRequest.Status and force reconcile if last synced version != current version for infrastructure resource. What do you think?

Reply:
Yes @akhil-rane, that looks to be a good idea. I would add the required changes and push the code.

Reply:
Thanks